src/panels/developer-tools/service/developer-tools-service.ts

```diff
@@ -353,19 +376,50 @@ class HaPanelDevService extends LitElement {
 
   private async _callService(ev) {
     const button = ev.currentTarget as HaProgressButton;
-    if (!this._serviceData?.service) {
+
+    if (this._yamlMode && !this._yamlValid) {
+      forwardHaptic("failure");
+      button.actionError();
+      showToast(this, {
+        message: this.hass.localize(
+          "ui.panel.developer-tools.tabs.services.errors.yaml.invalid_yaml"
+        ),
+      });
       return;
     }
-    const [domain, service] = this._serviceData.service.split(".", 2);
+
+    const { target, fields } = this._fields(
+      this.hass.services,
+      this._serviceData?.service
+    );
+
+    const isValid = this._isValid(
+      this._serviceData,
+      fields,
+      target,
+      this._yamlMode,
+      this.hass.localize
+    );
+
+    if (!isValid) {
+      forwardHaptic("failure");
+      button.actionError();
+      showToast(this, { message: this._error! });
+      return;
+    }
+    const [domain, service] = this._serviceData!.service!.split(".", 2);
     const script: Action[] = [];
-    if ("response" in this.hass.services[domain][service]) {
+    if (
+      this.hass.services?.[domain]?.[service] &&
+      "response" in this.hass.services[domain][service]
+    ) {
       script.push({
-        ...this._serviceData,
+        ...this._serviceData!,
```

Comment:
I'm not exactly sure why I need to `!` these serviceData.service vars here. `_isValid` seems to pretty clearly to me return false when it is undefined. Not sure if that's just a limitation of the linter, or if I've missed some subtle way these can be invalid.

Reply:
Yeah, it can't follow that